File: 00-coding-101/python/basics/control_flow.py
# Python Control Flow
# This file demonstrates control flow structures in Python

print("=== Python Control Flow ===\n")

# 1. If Statements
print("1. IF STATEMENTS:")
age = 18

if age >= 18:
    print("You are an adult")
elif age >= 13:
    print("You are a teenager")
else:
    print("You are a child")

# Multiple conditions
temperature = 25
humidity = 60

if temperature > 30 and humidity > 70:
    print("It's hot and humid!")
elif temperature > 25 or humidity > 50:
    print("It's warm or humid")
else:
    print("The weather is pleasant")
print()

# 2. For Loops
print("2. FOR LOOPS:")

# Loop through a list
fruits = ["apple", "banana", "orange", "grape"]
for fruit in fruits:
    print(f"I like {fruit}")

# Loop with range
print("\nCounting from 1 to 5:")
for i in range(1, 6):
    print(f"Count: {i}")

# Loop with enumerate
print("\nFruits with index:")
for index, fruit in enumerate(fruits):
    print(f"{index + 1}. {fruit}")

# Loop through dictionary
person = {"name": "Alice", "age": 25, "city": "New York"}
print("\nPerson details:")
for key, value in person.items():
    print(f"{key}: {value}")
print()

# 3. While Loops
print("3. WHILE LOOPS:")

# Simple while loop
count = 0
while count < 3:
    print(f"Count: {count}")
    count += 1

# While loop with break
print("\nWhile loop with break:")
number = 0
while True:
    if number >= 5:
        break
    print(f"Number: {number}")
    number += 1

# While loop with continue
print("\nWhile loop with continue:")
i = 0
while i < 10:
    i += 1
    if i % 2 == 0:  # Skip even numbers
        continue
    print(f"Odd number: {i}")
print()

# 4. List Comprehensions
print("4. LIST COMPREHENSIONS:")

# Basic list comprehension
numbers = [1, 2, 3, 4, 5]
squares = [x**2 for x in numbers]
print(f"Numbers: {numbers}")
print(f"Squares: {squares}")

# List comprehension with condition
even_squares = [x**2 for x in numbers if x % 2 == 0]
print(f"Even squares: {even_squares}")

# List comprehension with multiple variables
pairs = [(x, y) for x in range(3) for y in range(3)]
print(f"Pairs: {pairs}")
print()

# 5. Match Statements (Python 3.10+)
print("5. MATCH STATEMENTS (Python 3.10+):")

def analyze_command(command):
    match command.split():
        case ["quit"]:
            return "Goodbye!"
        case ["help"]:
            return "Available commands: quit, help, status"
        case ["status"]:
            return "System is running"
        case ["echo", *words]:
            return f"Echo: {' '.join(words)}"
        case _:
            return "Unknown command"

# Test match statements
commands = ["quit", "help", "status", "echo hello world", "unknown"]
for cmd in commands:
    result = analyze_command(cmd)
    print(f"'{cmd}' -> {result}")
print()

# 6. Error Handling
print("6. ERROR HANDLING:")

# Try-except block
try:
    number = int(input("Enter a number: "))
    result = 10 / number
    print(f"Result: {result}")
except ValueError:
    print("That's not a valid number!")
except ZeroDivisionError:
    print("Cannot divide by zero!")
except Exception as e:
    print(f"An error occurred: {e}")
else:
    print("No errors occurred!")
finally:
    print("This always runs")
print()

# 7. Nested Control Structures
print("7. NESTED CONTROL STRUCTURES:")

# Nested if statements
user_role = "admin"
user_age = 25

if user_role == "admin":
    if user_age >= 18:
        print("Full admin access granted")
    else:
        print("Limited admin access (under 18)")
elif user_role == "user":
    if user_age >= 18:
        print("Standard user access")
    else:
        print("Restricted access (under 18)")
else:
    print("Guest access only")

# Nested loops
print("\nMultiplication table (1-3):")
for i in range(1, 4):
    for j in range(1, 4):
        print(f"{i} x {j} = {i * j}", end="  ")
    print()  # New line after each row
print()

# 8. Practical Examples
print("8. PRACTICAL EXAMPLES:")

# Grade calculator
def calculate_grade(score):
    if score >= 90:
        return "A"
    elif score >= 80:
        return "B"
    elif score >= 70:
        return "C"
    elif score >= 60:
        return "D"
    else:
        return "F"

scores = [95, 87, 72, 55, 100]
for score in scores:
    grade = calculate_grade(score)
    print(f"Score: {score} -> Grade: {grade}")

# Password strength checker
def check_password_strength(password):
    has_letter = any(c.isalpha() for c in password)
    has_digit = any(c.isdigit() for c in password)
    has_special = any(c in "!@#$%^&*" for c in password)
    
    if len(password) >= 8 and has_letter and has_digit and has_special:
        return "Strong"
    elif len(password) >= 6 and has_letter and has_digit:
        return "Medium"
    else:
        return "Weak"

passwords = ["abc123", "password123!", "weak", "StrongPass1!"]
for pwd in passwords:
    strength = check_password_strength(pwd)
    print(f"Password '{pwd}': {strength}")

print("\n=== End of Control Flow ===") 

File: 00-coding-101/javascript/basics/control_flow.js
// JavaScript Control Flow
// This file demonstrates control flow structures in JavaScript

console.log("=== JavaScript Control Flow ===\n");

// 1. If Statements
console.log("1. IF STATEMENTS:");
let age = 18;

if (age >= 18) {
    console.log("You are an adult");
} else if (age >= 13) {
    console.log("You are a teenager");
} else {
    console.log("You are a child");
}

// Multiple conditions
let temperature = 25;
let humidity = 60;

if (temperature > 30 && humidity > 70) {
    console.log("It's hot and humid!");
} else if (temperature > 25 || humidity > 50) {
    console.log("It's warm or humid");
} else {
    console.log("The weather is pleasant");
}
console.log();

// 2. For Loops
console.log("2. FOR LOOPS:");

// Loop through an array
let fruits = ["apple", "banana", "orange", "grape"];
for (let fruit of fruits) {
    console.log(`I like ${fruit}`);
}

// Traditional for loop
console.log("\nCounting from 1 to 5:");
for (let i = 1; i <= 5; i++) {
    console.log(`Count: ${i}`);
}

// Loop with forEach
console.log("\nFruits with forEach:");
fruits.forEach((fruit, index) => {
    console.log(`${index + 1}. ${fruit}`);
});

// Loop through object
let person = { name: "Alice", age: 25, city: "New York" };
console.log("\nPerson details:");
for (let [key, value] of Object.entries(person)) {
    console.log(`${key}: ${value}`);
}
console.log();

// 3. While Loops
console.log("3. WHILE LOOPS:");

// Simple while loop
let count = 0;
while (count < 3) {
    console.log(`Count: ${count}`);
    count++;
}

// While loop with break
console.log("\nWhile loop with break:");
let number = 0;
while (true) {
    if (number >= 5) {
        break;
    }
    console.log(`Number: ${number}`);
    number++;
}

// While loop with continue
console.log("\nWhile loop with continue:");
let i = 0;
while (i < 10) {
    i++;
    if (i % 2 === 0) { // Skip even numbers
        continue;
    }
    console.log(`Odd number: ${i}`);
}
console.log();

// 4. Array Methods
console.log("4. ARRAY METHODS:");

let numbers = [1, 2, 3, 4, 5];

// Map - transform elements
let squares = numbers.map(x => x ** 2);
console.log(`Numbers: ${numbers}`);
console.log(`Squares: ${squares}`);

// Filter - select elements
let evenNumbers = numbers.filter(x => x % 2 === 0);
console.log(`Even numbers: ${evenNumbers}`);

// Reduce - combine elements
let sum = numbers.reduce((acc, curr) => acc + curr, 0);
console.log(`Sum: ${sum}`);

// Find - find first matching element
let firstEven = numbers.find(x => x % 2 === 0);
console.log(`First even number: ${firstEven}`);

// Some - check if any element matches
let hasEven = numbers.some(x => x % 2 === 0);
console.log(`Has even numbers: ${hasEven}`);

// Every - check if all elements match
let allPositive = numbers.every(x => x > 0);
console.log(`All positive: ${allPositive}`);
console.log();

// 5. Switch Statements
console.log("5. SWITCH STATEMENTS:");

function getDayName(dayNumber) {
    switch (dayNumber) {
        case 1:
            return "Monday";
        case 2:
            return "Tuesday";
        case 3:
            return "Wednesday";
        case 4:
            return "Thursday";
        case 5:
            return "Friday";
        case 6:
        case 7:
            return "Weekend";
        default:
            return "Invalid day";
    }
}

for (let day = 1; day <= 8; day++) {
    console.log(`Day ${day}: ${getDayName(day)}`);
}
console.log();

// 6. Error Handling
console.log("6. ERROR HANDLING:");

function divideNumbers(a, b) {
    try {
        if (typeof a !== 'number' || typeof b !== 'number') {
            throw new Error('Both arguments must be numbers');
        }
        if (b === 0) {
            throw new Error('Cannot divide by zero');
        }
        return a / b;
    } catch (error) {
        console.error(`Error: ${error.message}`);
        return null;
    } finally {
        console.log("Division operation completed");
    }
}

console.log(`10 / 2 = ${divideNumbers(10, 2)}`);
console.log(`10 / 0 = ${divideNumbers(10, 0)}`);
console.log(`10 / 'abc' = ${divideNumbers(10, 'abc')}`);
console.log();

// 7. Nested Control Structures
console.log("7. NESTED CONTROL STRUCTURES:");

// Nested if statements
let userRole = "admin";
let userAge = 25;

if (userRole === "admin") {
    if (userAge >= 18) {
        console.log("Full admin access granted");
    } else {
        console.log("Limited admin access (under 18)");
    }
} else if (userRole === "user") {
    if (userAge >= 18) {
        console.log("Standard user access");
    } else {
        console.log("Restricted access (under 18)");
    }
} else {
    console.log("Guest access only");
}

// Nested loops
console.log("\nMultiplication table (1-3):");
for (let i = 1; i <= 3; i++) {
    let row = "";
    for (let j = 1; j <= 3; j++) {
        row += `${i} x ${j} = ${i * j}  `;
    }
    console.log(row);
}
console.log();

// 8. Practical Examples
console.log("8. PRACTICAL EXAMPLES:");

// Grade calculator
function calculateGrade(score) {
    if (score >= 90) return "A";
    else if (score >= 80) return "B";
    else if (score >= 70) return "C";
    else if (score >= 60) return "D";
    else return "F";
}

let scores = [95, 87, 72, 55, 100];
scores.forEach(score => {
    let grade = calculateGrade(score);
    console.log(`Score: ${score} -> Grade: ${grade}`);
});

// Password strength checker
function checkPasswordStrength(password) {
    let hasLetter = /[a-zA-Z]/.test(password);
    let hasDigit = /\d/.test(password);
    let hasSpecial = /[!@#$%^&*]/.test(password);
    
    if (password.length >= 8 && hasLetter && hasDigit && hasSpecial) {
        return "Strong";
    } else if (password.length >= 6 && hasLetter && hasDigit) {
        return "Medium";
    } else {
        return "Weak";
    }
}

let passwords = ["abc123", "password123!", "weak", "StrongPass1!"];
passwords.forEach(pwd => {
    let strength = checkPasswordStrength(pwd);
    console.log(`Password '${pwd}': ${strength}`);
});

// 9. Modern JavaScript Features
console.log("\n9. MODERN JAVASCRIPT FEATURES:");

// Optional chaining
let user = {
    profile: {
        name: "John",
        address: {
            city: "New York"
        }
    }
};

console.log(`User city: ${user?.profile?.address?.city}`);
console.log(`User country: ${user?.profile?.address?.country || 'Unknown'}`);

// Nullish coalescing
let defaultValue = null;
let value = defaultValue ?? "Default Value";
console.log(`Value: ${value}`);

// Template literals with expressions
let name = "Alice";
let greeting = `Hello, ${name}! You are ${age} years old.`;
console.log(greeting);

console.log("\n=== End of Control Flow ==="); 